AI/scripts/util/slice_utils.py

```diff
@@ -0,0 +1,204 @@
+    return slice_coords_dict
+
+
+def perspective_transform(image, normalized_annotations, theta=0.7, gamma=0.3):
+    """
+    This is an implementation of perspective transform data augmentation as described in:
+    https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=8943416.
+
+    An image is transformed along with the annotations it contains.
+
+    :param image: a PIL image
+    :param normalized_annotations: the normalized annotations (x1, y1, x2, y2, label)
+    :param theta: the image plane rotation (maybe)
+    :param gamma: unknown TODO ?
+    :return: the transformed image and transformed normalized annotations
+    """
+
+    # image dimensions
+    width = image.width
+    height = image.height
+
+    # get coordinates for perspective
+    x_tl, x_bl, x_br, x_tr = \
+        [random.uniform(0, width * gamma * theta) for _ in range(4)]
+    y_tl, y_bl, y_br, y_tr = \
+        [random.uniform(0, height * gamma * theta) for _ in range(4)]
+
+    A_theta = [x_tl, y_tl]  # top left
+    B_theta = [x_bl, height - y_bl]  # bottom left
+    C_theta = [width - x_br, height - y_br]  # bottom right
+    D_theta = [width - x_tr, y_tr]  # top right
+
+    # create transformation matrix
+    pts1 = np.float32([[0, 0], [0, height], [width, height], [width, 0]])
+    pts2 = np.float32([A_theta, B_theta, C_theta, D_theta])
+    matrix = cv2.getPerspectiveTransform(pts1, pts2)
+
+    # iterate over all annotations and apply the perspective transform
+    transformed_annotations = list()
+    for x1, y1, x2, y2, label in normalized_annotations:
+        # coordinates for all four corners of annotation; must use absolute coordinates
+        x_tl, x_bl, x_tr, x_br = np.array((x1, x1, x2, x2)) * width
+        y_tl, y_bl, y_tr, y_br = np.array((y1, y1, y2, y2)) * height
+
+        # transform each coordinate
+        pts = zip((x_tl, x_bl, x_tr, x_br), (y_tl, y_bl, y_tr, y_br))
+        p_tl, p_bl, p_tr, p_br = \
+            np.squeeze(
+                cv2.perspectiveTransform(
+                    np.array(
+                        [[[x, y]] for x, y in pts], dtype='float32'
+                    ),
+                    matrix
+                ))
+
+        # get coordinates of new bounding box based on extrema; must convert back to normalized coordinates
+        x_min = min(p_tl[0], p_bl[0], p_tr[0], p_br[0]) / width
+        x_max = max(p_tl[0], p_bl[0], p_tr[0], p_br[0]) / width
+        y_min = min(p_tl[1], p_bl[1], p_tr[1], p_br[1]) / height
+        y_max = max(p_tl[1], p_bl[1], p_tr[1], p_br[1]) / height
+
+        # collect the new annotation
+        transformed_annotations.append((x_min, y_min, x_max, y_max, label))
+
+    # apply the perspective transform to the image; must convert to OpenCV image first
+    open_cv_image = cv2.cvtColor(np.array(image), cv2.COLOR_RGBA2mRGBA)
+    result = cv2.warpPerspective(open_cv_image, matrix, (width, height))
+
+    # TODO transform results in transparent borders.
```

Comment:
what does this mean?

Reply:
The image gets warped, so it may no longer oriented the same, and it likely no longer has 90 degree corners. Since I do not crop it, the warpPerspective fills the empty space with transparency. The paper I based this off of has experimented with several options for this, including cropping, reflection, and filling the space with black. This is something we may want to revisit based on performance.